src/content_main.js

```diff
@@ -61,6 +68,88 @@ function setExtensionIconMonetizationState(isMonetized) {
 	webBrowser.runtime.sendMessage({ isMonetized });
 }
 
+/***********************************************************
+ * Track Visits and Time Spent on Website
+ ***********************************************************/
+
+/**
+ * Track the current visit to the site (origin).
+ * No data needs to be passed in, since it is handled in AkitaOriginVisitData.
+ */
+async function trackVisitToSite() {
+	await storeDataIntoAkitaFormat(null, AKITA_DATA_TYPE.ORIGIN_VISIT_DATA);
+}
+
+/**
+ * Calculate and store the time the user spends on the site.
+ * Call this function once at the beginning of website logic.
+ * 
+ * Use the Page Visibility API to check if the current webpage is visible or not.
+ * https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API
+ * https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState
+ */
+async function trackTimeOnSite() {
+	let previousStoreTime = getCurrentTime();
+	let docHiddenTime = -1;
+	let docVisibleTime = -1;
+
+	document.addEventListener('visibilitychange', (event) => {
+		if (document.hidden) {
+			// The page is no longer visible
+			docHiddenTime = getCurrentTime();
+		} else {
+			// The page is now visible
+			docVisibleTime = getCurrentTime();
+		}
+	});
+
+	/**
+	 * NOTE: setInterval may not be called while the document is hidden (while visibility lost)
+	 * https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Reasons_for_delays_longer_than_specified
+	 * 
+	 * Store the recent time spent every 2 seconds to ensure time spent on site is recorded
+	 * even if the user closes the site.
+	 */
+	setInterval(async () => {
+		const now = getCurrentTime();
+		
+		if (document.visibilityState === 'visible') {
+			if (docHiddenTime > previousStoreTime) {
+				// Adding time after visibility lost (document becomes hidden) and then gained (document is visible again)
+				// i.e. If the user navigates away from the site and then comes back
+				const timeFromPreviousStoreToDocHidden = docHiddenTime - previousStoreTime;
+				const timeSinceDocVisible = now - docVisibleTime;	
+				await storeRecentTimeSpent(timeFromPreviousStoreToDocHidden + timeSinceDocVisible);
+			} else {
+				// Adding time during regular interval (document visible)
+				await storeRecentTimeSpent(now - previousStoreTime);
+			}
+
+			previousStoreTime = now;
+		}
+	}, 2000); // 2 second interval
+}
+
+/**
+ * Get the current time based on the user's timezone.
+ */
+function getCurrentTime() {
+	return new Date();
```

Comment:
FYI: a more accurate alternative to `new Date()` for the purposes of timing is `performance.now()`.

Reply:
Just a note, `new Date()` and `performance.now()` have reduced time precision to protect against attacks.

`new Date()`
> In Firefox, the privacy.reduceTimerPrecision  preference is enabled by default and defaults to 20µs in Firefox 59; in 60 it will be 2ms.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now#Reduced_time_precision

`performance.now()`
> In Firefox, the privacy.reduceTimerPrecision preference is enabled by default and defaults to 1ms.

https://developer.mozilla.org/en-US/docs/Web/API/Performance/now#Reduced_time_precision

---

Although `performance.now()` has more precision than `new Date()`, it might be less performant as a result - although its performance seems to be highly dependent on browser. As you can see below, the performance is pretty opposite.

**Chrome**
![image](https://user-images.githubusercontent.com/25834218/83705451-faf87400-a5e2-11ea-8be7-56dacbdcab92.png)

**Firefox**
![image](https://user-images.githubusercontent.com/25834218/83705475-0a77bd00-a5e3-11ea-8b80-54710e7f8f52.png)

Reply:
Informative! Nice.